src/lib/pubkey/classic_mceliece/cmce_encaps.cpp

```diff
@@ -0,0 +1,126 @@
+      // This side channel only leaks which random elements are selected and which are dropped,
+      // but no information about their content is leaked.
+      d &= mask_m;
+      if(d < params.n() && a_values.size() < params.t()) {
+         a_values.push_back(d);
+      }
+   }
+   if(a_values.size() < params.t()) {
+      // Step 3: ... If fewer than t of such elements exist restart
+      return std::nullopt;
+   }
+
+   // Step 4: Restart if not all a_i are distinct
+   for(size_t i = 1; i < params.t(); ++i) {
+      for(size_t j = 0; j < i; ++j) {
+         if(a_values.at(i) == a_values.at(j)) {
+            return std::nullopt;
+         }
+      }
+   }
+
+   secure_vector<uint8_t> a_value_byte(params.t());
+   secure_vector<uint8_t> e_bytes(ceil_tobytes(params.n()));
+
+   // Step 5: Set all bits of e at the positions of a_values
+   // Prepare the associated byte in e_bytes that is represented by each bit index in a_values
+   // if we e is represented as a byte vector
+   for(size_t j = 0; j < a_values.size(); ++j) {
+      a_value_byte[j] = 1 << (a_values[j] % 8);
+   }
+
+   for(size_t i = 0; i < params.n() / 8; ++i) {
+      for(size_t j = 0; j < a_values.size(); ++j) {
+         // If the current byte is the one that is represented by the current bit index in a_values
+         // then set the bit in e_bytes (in-byte position prepared above)
+         auto mask = CT::Mask<uint16_t>::is_equal(static_cast<uint16_t>(i), static_cast<uint16_t>(a_values[j] >> 3));
+         e_bytes[i] |= mask.if_set_return(a_value_byte[j]);
+      }
+   }
+
+   return secure_bitvector(e_bytes, params.n());
+}
+
+void Classic_McEliece_Encryptor::raw_kem_encrypt(std::span<uint8_t> out_encapsulated_key,
+                                                 std::span<uint8_t> out_shared_key,
+                                                 RandomNumberGenerator& rng) {
+   BOTAN_ASSERT(out_encapsulated_key.size() == m_key->params().ciphertext_size(),
+                "Correct encapsulated key output length");
+   BOTAN_ASSERT(out_shared_key.size() == m_key->params().hash_out_bytes(), "Correct shared key output length");
+
+   auto& params = m_key->params();
+
+   // Call fixed_weight until it is successful to
+   // create a random error vector e of weight tau
+   secure_bitvector e = [&] {
+      // Emergency abort in case unexpected logical error to prevent endless loops
+      //   Success probability: >24% per attempt (25% that elements are distinct * 96% enough elements are in range)
+      //   => 203 attempts for 2^(-80) fail probability
+      const size_t MAX_ATTEMPTS = 203;
+      for(size_t attempt = 0; attempt < MAX_ATTEMPTS; ++attempt) {
+         if(auto maybe_e = fixed_weight_vector_gen(params, rng.random_vec((params.sigma1() / 8) * params.tau()))) {
+            return maybe_e.value();
+         }
+      }
+      throw Internal_Error("Cannot created fixed weight vector.");
```

Comment:
```suggestion
      throw Internal_Error("Cannot create fixed weight vector. Is your RNG broken?");
```

..also: Do we have a test for this? I.e. with a spiked RNG, that mocks a broken entropy source

Reply:
I added a test case (see c218cdd).